commit ebb620d9e5f79d90701f70bff5dbe7432b2db883
Author: Marta Lofstedt <marta.lofstedt@intel.com>
Date:   Mon Dec 5 14:04:08 2016 +0200

    drm: allow changing DPMS mode
    
    The drm_atomic_helper_connector_dpms
    will set the connector back the old DPMS state
    before returning. This makes it impossible to change
    DPMS state of the connector.
    
    Fixes: 0853695c3ba46f97dfc0b5885f7b7e640ca212dd
    v2: edit of commit message
    Cc: Chris Wilson <chris@chris-wilson.co.uk>
    Cc: Daniel Vetter <daniel.vetter@ffwll.ch>
    Cc: Eric Engestrom <eric.engestrom@imgtec.com>
    Cc: Sean Paul <seanpaul@chromium.org>
    Cc: dri-devel@lists.freedesktop.org
    Cc: <drm-intel-fixes@lists.freedesktop.org>
    Signed-off-by: Marta Lofstedt <marta.lofstedt@intel.com>
    Signed-off-by: Daniel Vetter <daniel.vetter@ffwll.ch>
    Link: http://patchwork.freedesktop.org/patch/msgid/20161205120408.13056-1-marta.lofstedt@intel.com
    (cherry picked from commit 8f5040e421ca4bbd8cd5216dec509da188eb74bc)

diff --git a/drivers/gpu/drm/drm_atomic_helper.c b/drivers/gpu/drm/drm_atomic_helper.c
index 50acd799babe..06572ffdb408 100644
--- a/drivers/gpu/drm/drm_atomic_helper.c
+++ b/drivers/gpu/drm/drm_atomic_helper.c
@@ -2932,7 +2932,8 @@ int drm_atomic_helper_connector_dpms(struct drm_connector *connector,
 	if (ret == -EDEADLK)
 		goto backoff;
 
-	connector->dpms = old_mode;
+	if (ret != 0)
+		connector->dpms = old_mode;
 	drm_atomic_state_free(state);
 
 	return ret;
